src/awkward/_nplikes/typetracer.py

```diff
@@ -249,7 +249,6 @@ def nplike(self) -> TypeTracer:
 
     @property
     def ndim(self) -> int:
-        self.touch_shape()
```

Comment:
The dimensionality is _not_ the shape, it's the length of the shape tuple. I think we need to treat this as a must-be-knowable.

Reply:
Yes, absolutely. There are three levels of knowledge about array metadata that you can have:

  * no information about `ndim`, like NumPy at compile-time (i.e. when compiling NumPy)
  * knowledge of `ndim` but not the values in `shape`, like a Numba function when JIT-compiling it
  * full knowledge of the `shape`, like JAX when JIT-compiling for XLA

It is absolutely reasonable for our model of arrays at Dask-DAG collection time (which is like a kind of compilation) to be like Numba's model. Knowing `ndim`, you know how many times you can `__getitem__(int)` before the type changes from an array to a scalar, for instance. That seems pretty fundamental.